lib/CapabilityDelegation.js

```diff
@@ -0,0 +1,122 @@
+/**
+ * Linked Data Signatures/Proofs
+ *
+ * @author Christopher Lemmer Webber
+ * @author Ganesh Annan <gannan@digitalbazaar.com>
+ * 
+ * @license BSD 3-Clause License
+ * Copyright (c) 2018 Digital Bazaar, Inc.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution.
+ *
+ * Neither the name of the Digital Bazaar, Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+ * IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+ * PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+ * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+'use strict';
+
+const jsigs = require('jsonld-signatures');
+const utils = require('./utils');
+const vocab = require('./vocab');
+
+const {ProofPurpose} = jsigs;
+
+// TODO: Maybe convert this to a non-recursive version that iterates through
+//   the cap chain as an array instead
+module.exports = class CapabilityDelegation extends ProofPurpose {
+  constructor(injector) {
+    super(injector);
+    this.documentLoader = injector.use('jsonld').documentLoader;
+    this.uri = vocab.capabilityDelegationUri;
+  }
+
+  // Arguments:
+  //  - document: An already-expanded version of the delegation document
+  //    this proof is attached to
+  //  - proof: An already-expanded version of the proof we are checking
+  //
+  // proofPurposeOptions keywords:
+  //  - expectedTarget: the target we expect this capability to apply to (URI).
+  //  - caveatVerifiers: a hashmap of caveat URIs to procedures which
+  //    verify those caveats.  The caveat-verifying procedures should take
+  //    two arguments: the caveat being verified, and the expanded invocation
+  //    document.
+  async verify(document, proof, proofPurposeOptions) {
```

Comment:
I can supply the reason for every "return false" if that helps

Reply:
I think the pattern followed in other verify/validate methods can apply here. Which is, instead of returning true/false, instead raise errors on false, and just return with noop when true. And the errors can provide explanation etc.

Reply:
@dmitrizagidulin, I think we should always be checking the return value for `true` or `false` or we're going to get into trouble. It's fine to throw when there are exceptions too, of course, but let's not encourage NOT checking the return value because this has already bitten us before.

Reply:
@dlongley Not sure I understand. Since the return value will never be `false`, why check for it?

Reply:
@dmitrizagidulin Why do you believe the return value will never be `false`?